os/drivers/compression/compress.c

```diff
@@ -116,10 +130,74 @@ static int comp_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
 		ret = OK;
 		break;
  	case COMPIOC_DECOMPRESS:
+		if(comp_info == NULL) {
+			return -EINVAL;
+		}
 		ret = decompress_block(comp_info->output_buffer, &comp_info->output_size, comp_info->input_buffer, &comp_info->input_size);
 		if (ret == ENOMEM) {
 			bcmpdbg("Output buffer allocated is not sufficient\n");
 		}
+#ifdef CONFIG_COMPRESSED_BINARY
+	case COMPIOC_FDCMP_INIT:
+		if((char *)arg == NULL) {
+			return -EINVAL;
+		}
+		data = (struct file_decomp_data_s *)malloc(sizeof(struct file_decomp_data_s));
+		data->fd = open((char *)arg, "r");
+		filep->f_priv = data;
+		if (data->fd < 0) {
+			berr("Failed to open file = %s\n", (char *)arg);
+			free(data->compression_header);
```

Comment:
We are not allocating the memory for compression_header. It is allocated and freed inside the compression code. So, we must never free this. Please fix this in all the places below also.